ipa-core/src/ff/prime_field.rs

```diff
@@ -454,10 +478,40 @@ mod fp61bit {
         pub fn from_bit(input: bool) -> Self {
             Self(input.into())
         }
+
+        #[must_use]
+        fn modulo_prime_base(val: u128) -> Self {
+            Self::modulo_prime_u128(val)
+        }
+
+        /// Implements optimized modulus division operation for Mersenne fields.
+        /// Implementation taken from [`bit_twiddling`].
+        ///
+        /// [`bit_twiddling`]: https://graphics.stanford.edu/~seander/bithacks.html#ModulusDivision
+        #[must_use]
+        #[allow(clippy::cast_possible_truncation)]
+        fn modulo_prime_u128(val: u128) -> Self
+        where
+            Self: U128Conversions,
+        {
+            const PRIME: u128 = Fp61BitPrime::PRIME as u128;
+            debug_assert_eq!(0, PRIME & (PRIME + 1), "{PRIME} is not a Mersenne prime");
+
+            let val = (val & PRIME) + (val >> Self::BITS);
+            // another round if val ended up being greater than PRIME
+            let val = (val & PRIME) + (val >> Self::BITS);
+            if val == PRIME {
+                Self::ZERO
```

Comment:
It may be possible to save a little more if we can reduce some of the operations here from `u128` to `u64`. I guess to do that we may need to figure out how to get rustc/llvm to infer an add-with-carry.

The other win will be if we can get `LagrangeTable::eval` (and possibly other places) to do several multiplies before each modulo reduction.

Reply:
lets talk about it in our next 1:1 - sounds very promising